### PR TITLE
Allow to find all references of the 'this 'keyword

### LINKED ISSUE #9037: this-function parameter can be renamed
```ts
function(this/**/: string, x: number) {
  return this.length;
}
```
Now try to rename or find all references at `/**/`.

Expected: Message "You cannot rename this element".

Actual: Rename *just* the parameter, nothing in the body. Find all references just finds one reference, the definition.

Probably `this` needs to be classified as ThisKeyword by the parser as long as that doesn't break anything else.

### LINKED ISSUE #9198: Find all refs and goto def do not work on `this` in function parameter position
```ts
function f(this:any) {
    this;// find all ref/ goto def here
}
```

I expect this should show the two references of `this`.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #9198. Closes #9037 because we don't need to prevent renaming if it works.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -3198,6 +3198,13 @@ namespace ts {
             return undefined;
         }
 
+        //neater
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: ?

Comment (andy-ms) [inline src/compiler/checker.ts]: Whoops, pushed WIP

File: src/compiler/types.ts [hunk 2]
```diff
@@ -2386,6 +2386,8 @@ namespace ts {
         declaration: SignatureDeclaration;  // Originating declaration
         typeParameters: TypeParameter[];    // Type parameters (undefined if non-generic)
         parameters: Symbol[];               // Parameters
+        //TODO: maybe this should point directly to the Symbol?
+        thisParameter: Symbol,
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: Mark this `/*@internal*/`, this is a temporary workaround.

**Comment by andy-ms** [comment]
Closed by #9220

File: src/compiler/types.ts [hunk 3]
```diff
@@ -2388,6 +2388,8 @@ namespace ts {
         parameters: Symbol[];               // Parameters
         thisType?: Type;                    // type of this-type
         /* @internal */
+        thisParameter?: Symbol;             // symbol of this-type parameter
```
Comment (sandersn) [inline src/compiler/types.ts]: It was never quite right to resolve `thisType` in getSignatureOfDeclaration` so switching to a symbol is an improvement. But it means that you should remove `thisType` entirely and replace usages with `getTypeOfSymbol(sig.thisParameter)`.

File: tests/cases/fourslash/findAllRefsThisKeyword.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+////    }
+////    method() {
+////        () => [|this|];
+////    }
+////}
+////// These are *not* real uses of the 'this' keyword, they are identifiers.
+////const x = { [|this|]: 0 }
+////x.[|this|];
+
+const [global, f0, f1, g0, g1, x, y, constructor, method, propDef, propUse] = test.ranges();
```
Comment (sandersn) [inline tests/cases/fourslash/findAllRefsThisKeyword.ts]: nice usage of destructuring to tame `ranges`!

**Comment by sandersn** [comment]
Looks good except for getting rid of `thisType` in favor of `thisParameter`.

**Comment by sandersn** [comment]
@yuit and I discussed the removal of `thisType` a little. `getTypeOfVariableOfParameterOrProperty` (called by `getTypeOfSymbol`) may need a new case, something like

```ts
if (declaration.kind === SyntaxKind.Identifier && declaration.originalKeyword.kind === SytaxKind.ThisKeyword) {
  return links.type = declaration.type ? getTypeFromTypeNode(declaration.type) : unknownType;
}
```

Of course you may need to add some type assertions in there.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -3198,14 +3197,13 @@ namespace ts {
...
-                    return getTypeFromTypeNode(accessor.parameters[0].type);
-                }
-            }
-            return undefined;
+        function getAnnotatedAccessorThisParameter(accessor: AccessorDeclaration): Symbol | undefined {
+            const parameter = getAccessorThisParameter(accessor);
+            return parameter && parameter.symbol;
+        }
+
+        function getAnnotatedAccessorThisType(accessor: AccessorDeclaration): Type | undefined {
```
Comment (sandersn) [inline src/compiler/checker.ts]: is this used now?

Comment (andy-ms) [inline src/compiler/checker.ts]: Linter would complain if it weren't.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -8594,9 +8602,10 @@ namespace ts {
                 if (type) {
                     return type;
                 }
-                const signature = getSignatureFromDeclaration(container);
-                if (signature.thisType) {
-                    return signature.thisType;
+
+                const thisType = getThisTypeOfSignature(getSignatureFromDeclaration(container));
```
Comment (sandersn) [inline src/compiler/checker.ts]: except for the argument type, this could be a call to `getAnnotatedAccessorThisType`

Comment (sandersn) [inline src/compiler/checker.ts]: Which makes me think it needs a more general name

File: src/compiler/checker.ts (hunk 5 above)
Comment (sandersn) [inline src/compiler/checker.ts]: I looked it up: `checkAccessorDeclaration` still uses it.

**Comment by sandersn** [comment]
:+1:

**Pull request merged**